File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;


class Solution {
    public HashSet<Integer> ans = new HashSet<>();
    
    public int solution(String numbers) {
        int n = numbers.length();
        
        int[] numberInts = new int[n];
        for (int idx = 0; idx < n; idx++)
            numberInts[idx] = numbers.charAt(idx) - '0';
            
        backtracking(numberInts, n, new boolean[n], 0);    
        return ans.size();
    }
    
    public void backtracking(int[] numberInts, int n, boolean[] visited, int make) {
        if (isPrime(make))
            ans.add(make);
        
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                visited[i] = true;
                backtracking(numberInts, n, visited, make*10 + numberInts[i]);
                visited[i] = false;
            }
        }
    }
    
    public boolean isPrime(int n) {
        if (n == 0 || n == 1)
            return false;
        for (int i = 2; i * i <= n; i++) {
            if (n % i == 0)
                return false;
        }
        return true;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from math import sqrt

s = set()
def solution(numbers):
    backtracking(numbers, 0, "0")
    return len(s)
    
    
def backtracking(numbers, used, make):
    if is_prime(int(make)):
        s.add(int(make))
        
    for idx, val in enumerate(numbers):
        idx_bit = 1 << idx
        if not idx_bit & used:
            backtracking(numbers, used | idx_bit, make + val)
        
    
def is_prime(n):
    if n in (0, 1):
        return False
    
    for i in range(2, int(sqrt(n))+1):
        if not n%i:
            return False
    return True
    